### PR TITLE
Make signs of fields exported by ATM consistent with positive-down for CMEPS

### LINKED ISSUE #93: CMEPS is forced to use custom merges because sign of field imported from ATM is wrong or unit conversion is required
CMEPS has the capability to define generic merges between components. However, these generic merges are not available when the sign of the field needs to change, or the field needs to be converted to one of different units (ie, latent->evap). 

Because of this issue, multiple "custom" merges are required in ``med_phases_prep_ocn``.  For example, the sensible heat flux from the ATM has the wrong sign, requiring a custom merge:

```
       customwgt(:) = -ofrac(:)
       call med_merge_field(is_local%wrap%FBExp(compocn),      'Faxa_sen',  &
            FBinA=is_local%wrap%FBImp(compatm,compocn), fnameA='Faxa_sen', wgtA=customwgt, rc=rc)
       if (ChkErr(rc,__LINE__,u_FILE_u)) return
```

The ATM exports the latent heat flux, but the ocean requires the evaporative flux, which requires a custom merge to change the sign and units of the field:

```
       customwgt(:) = -ofrac(:) / const_lhvap
       call med_merge_field(is_local%wrap%FBExp(compocn),      'Faxa_evap', &
            FBinA=is_local%wrap%FBImp(compatm,compocn), fnameA='Faxa_lat' , wgtA=customwgt, rc=rc)
       if (ChkErr(rc,__LINE__,u_FILE_u)) return
```

The momemtum fluxes from the ATM also require a custom merge to change the sign. 

This adds needless complexity to maintain these custom merges when a better solution is to have the ATM export the coupling field with the right sign or unit. This also makes it possible to leverage the existing diagnostic capabilities with CMEPS to examine global budgets for UFS applications.

### LINKED ISSUE #94: remove aoflux calculation when using coupling_mode='hafs'
The hafs coupling mode does not use aoflux calculation but the current code sets up merges and field bundles as though it is used. This adds unneeded and confusing code. The aoflux options under 'hafs' should be removed. 

- related to #93

### LINKED ISSUE #669: ATM needs to export coupling fields to mediator with correct sign or unit
## Description
Currently specialized code is required in CMEPS because the export fields obtained from the ATM do not have the correct sign or unit. For example, MOM6 requires the evaporative flux, not the latent heat flux, and so specialized code is required in CMEPS just to change the sign and unit.  In addition, the momementum stresses and sensible heat flux also need a sign change within CMEPS. See https://github.com/NOAA-EMC/CMEPS/issues/93

## Solution
In ``atmos_model.F90``, both sign and units can be changed using the optional ``scale_factor`` argument. For example,

```
	    ! MEAN Zonal compt of momentum flux (N/m**2)
            case ('mean_zonal_moment_flx_atm')
              call block_data_copy(datar82d, GFS_data(nb)%coupling%dusfc_cpl, Atm_block, nb, scale_factor=-rtime, rc=localrc)
```

## Alternatives
Maintain existing code

## Testing:

I have a working [branch](https://github.com/DeniseWorthen/fv3atm/tree/feature/signs_units) which is B4B with current baselines with the exception of the mediator restart files (ie, the data is the same, but with an opposite sign).

However, this method does change the value of the "non-valid" value in the exported fields. Currently, these non-valid values are typically ``1.385E+34``.  An optional masking value could be used in the block_data interfaces to keep the values the same. I need to know where this value of ``1.385E+34`` is set however.

### PR DESCRIPTION
**Pull request by DeniseWorthen**
## Description

Current special merges in CMEPS and/or sign changes in component caps are required because the ATM component exports fields with the wrong sign (from the perspective of the component) and/or with because the ATM is exporting latent heat when the component expects the evaporative flux. 

The fields affected are the momentum stresses and the sensible heat flux, both of which have the incorrect sign for MOM6 and/or HYCOM and the latent heat flux, which must be converted to evaporative flux for MOM6 within CMEPS.

Note: The change to allow ATM to export evap flux was merged previously in PR #1812 

This PR applies the change of sign and/or unit conversion when the field is exported by the ATM and removes the specialized conversions and/or merges in CMEPS or the relevant cap. The exported fields will be consistent w/ the coupling definition of "positive down". This will make it easier to extend both the use of the diagnostic capabilities of CMEPS as well refactoring the ``esmFldsExhange_nems`` to remove much of the specialized logic for merges depending on the ``coupling_mode`` (ie, nems_orig, nems_orig_data etc).

### Input data additions/changes
- [X] No changes are expected to input data.
- [ ] Changes are expected to input data:
  - [ ] New input data.
  - [ ] Updated input data.

### Anticipated changes to regression tests:
- [ ] No changes are expected to any regression test.
- [X] Changes are expected to the following tests:

Cpld and hafs test change baselines because sign for the sensible, evaporative, latent and momentum within the mediator files are different. This was verified by running the full RT was run on Hera at commit f767818d. In this case, all files except mediator history and restart files reproduce. Note that all DATM (``ng-godas``) tests reproduce because the fluxes in those cases are computed in the mediator and have the correct sense of positive down.

[RegressionTests_hera.log](https://github.com/ufs-community/ufs-weather-model/files/11986358/RegressionTests_hera.log)
<details><summary>Failed tests</summary>

```sh

7399:001 cpld_control_p8_mixedmode_intel failed in check_result
7401:002 cpld_control_gfsv17_intel failed in check_result
7403:003 cpld_control_p8_intel failed in check_result
7405:005 cpld_control_qr_p8_intel failed in check_result
7407:007 cpld_2threads_p8_intel failed in check_result
7409:008 cpld_decomp_p8_intel failed in check_result
7411:009 cpld_mpi_p8_intel failed in check_result
7413:010 cpld_control_ciceC_p8_intel failed in check_result
7415:011 cpld_control_c192_p8_intel failed in check_result
7417:013 cpld_bmark_p8_intel failed in check_result
7419:015 cpld_control_noaero_p8_intel failed in check_result
7421:016 cpld_control_nowave_noaero_p8_intel failed in check_result
7423:017 cpld_debug_p8_intel failed in check_result
7425:018 cpld_debug_noaero_p8_intel failed in check_result
7427:020 cpld_control_c48_intel failed in check_result
7429:021 cpld_control_p8_faster_intel failed in check_result
7431:022 cpld_control_pdlib_p8_intel failed in check_result
7433:025 cpld_debug_pdlib_p8_intel failed in check_result
7435:132 hafs_regional_atm_ocn_intel failed in check_result
7437:134 hafs_regional_atm_ocn_wav_intel failed in check_result
7439:149 hafs_regional_docn_intel failed in check_result
7441:150 hafs_regional_docn_oisst_intel failed in check_result
7443:151 hafs_regional_datm_cdeps_intel failed in check_result
7445:232 cpld_control_p8_gnu failed in check_result
7447:233 cpld_control_nowave_noaero_p8_gnu failed in check_result
7449:234 cpld_debug_p8_gnu failed in check_result
7451:235 cpld_control_pdlib_p8_gnu failed in check_result
7453:236 cpld_debug_pdlib_p8_gnu failed in check_result
```
</details>

<details><summary>Non-reproducing files</summary>

```sh
tests/logs/log_hera: grep NOT rt_* 

rt_001_cpld_control_p8_mixedmode_intel.log:64: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_002_cpld_control_gfsv17_intel.log:63: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_003_cpld_control_p8_intel.log:64: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_005_cpld_control_qr_p8_intel.log:64: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_007_cpld_2threads_p8_intel.log:52: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_008_cpld_decomp_p8_intel.log:52: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_009_cpld_mpi_p8_intel.log:52: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_010_cpld_control_ciceC_p8_intel.log:64: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_011_cpld_control_c192_p8_intel.log:52: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-43200.nc ............ALT CHECK......NOT OK
rt_013_cpld_bmark_p8_intel.log:47: Comparing RESTART/ufs.cpld.cpl.r.2013-04-01-21600.nc ............ALT CHECK......NOT OK
rt_015_cpld_control_noaero_p8_intel.log:63: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_016_cpld_control_nowave_noaero_p8_intel.log:63: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_017_cpld_debug_p8_intel.log:52: Comparing RESTART/ufs.cpld.cpl.r.2021-03-22-32400.nc ............ALT CHECK......NOT OK
rt_018_cpld_debug_noaero_p8_intel.log:51: Comparing RESTART/ufs.cpld.cpl.r.2021-03-22-32400.nc ............ALT CHECK......NOT OK
rt_020_cpld_control_c48_intel.log:51: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_021_cpld_control_p8_faster_intel.log:64: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_022_cpld_control_pdlib_p8_intel.log:63: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_025_cpld_debug_pdlib_p8_intel.log:51: Comparing RESTART/ufs.cpld.cpl.r.2021-03-22-32400.nc ............ALT CHECK......NOT OK
rt_132_hafs_regional_atm_ocn_intel.log:9: Comparing ufs.hafs.cpl.hi.2019-08-29-21600.nc ............ALT CHECK......NOT OK
rt_132_hafs_regional_atm_ocn_intel.log:10: Comparing ufs.hafs.cpl.r.2019-08-29-21600.nc ............ALT CHECK......NOT OK
rt_134_hafs_regional_atm_ocn_wav_intel.log:12: Comparing ufs.hafs.cpl.r.2019-08-29-21600.nc ............ALT CHECK......NOT OK
rt_149_hafs_regional_docn_intel.log:7: Comparing ufs.hafs.cpl.hi.2019-08-29-21600.nc ............ALT CHECK......NOT OK
rt_149_hafs_regional_docn_intel.log:8: Comparing ufs.hafs.cpl.r.2019-08-29-21600.nc ............ALT CHECK......NOT OK
rt_150_hafs_regional_docn_oisst_intel.log:7: Comparing ufs.hafs.cpl.hi.2019-08-29-21600.nc ............ALT CHECK......NOT OK
rt_150_hafs_regional_docn_oisst_intel.log:8: Comparing ufs.hafs.cpl.r.2019-08-29-21600.nc ............ALT CHECK......NOT OK
rt_151_hafs_regional_datm_cdeps_intel.log:5: Comparing ufs.hafs.cpl.hi.2019-08-30-00000.nc ............ALT CHECK......NOT OK
rt_151_hafs_regional_datm_cdeps_intel.log:6: Comparing ufs.hafs.cpl.r.2019-08-30-00000.nc ............ALT CHECK......NOT OK
rt_232_cpld_control_p8_gnu.log:64: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_233_cpld_control_nowave_noaero_p8_gnu.log:63: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_234_cpld_debug_p8_gnu.log:52: Comparing RESTART/ufs.cpld.cpl.r.2021-03-22-32400.nc ............ALT CHECK......NOT OK
rt_235_cpld_control_pdlib_p8_gnu.log:63: Comparing RESTART/ufs.cpld.cpl.r.2021-03-23-21600.nc ............ALT CHECK......NOT OK
rt_236_cpld_debug_pdlib_p8_gnu.log:51: Comparing RESTART/ufs.cpld.cpl.r.2021-03-22-32400.nc ............ALT CHECK......NOT OK
  ```
  </details>

## Subcomponents involved:
- [ ] AQM
- [ ] CDEPS
- [ ] CICE
- [X] CMEPS
- [ ] CMakeModules
- [X] FV3
- [ ] GOCART
- [X] HYCOM
- [ ] MOM6
- [ ] NOAHMP
- [ ] WW3
- [ ] stochastic_physics
- [ ] none

### Library Updates/Changes

- [x] Not Needed
- [ ] Create separate issue in [JCSDA/spack-stack](https://github.com/JCSDA/spack-stack) asking for update to library. Include library name, library version.
- [ ] Add issue link from JCSDA/spack-stack following this item

## Commit Queue Checklist:

- [x] Link PR's from all sub-components involved in section below
- [x] Confirm reviews completed in ALL sub-component PR's
- [x] Add all appropriate labels to this PR.
- [x] Run full RT suite on either Hera/Cheyenne AND attach log to a PR comment.
- [x] Add list of any failed regression tests to "Anticipated changes to regression tests" section.

## Linked PR's and Issues:

- fixes https://github.com/NOAA-EMC/fv3atm/issues/669
- fixes https://github.com/NOAA-EMC/CMEPS/issues/93
- fixes https://github.com/NOAA-EMC/CMEPS/issues/94

- [ ] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR.
- [ ] Move new/updated input data on RDHPCS Hera and propagate input data changes to all supported systems.

### Testing Log (for CM's):
- RDHPCS
  - [x] Hera
  - [x] Orion
  - [x] Jet
  - [x] Gaea
  - [x] Cheyenne
- WCOSS2
  - [x] Dogwood/Cactus
  - [x] Acorn
- CI
  - [x] Completed
- opnReqTest
  - [ ] N/A
  - [x] Log attached to comment

### CODE REVIEW

**Comment by junwang-noaa** [comment]
Is the sign swapping in HYCOM required just because the sign in swapped in fv3? I still feel this rule of requiring all the components to provide "positive down" in CMEPS could add additional burden in the component level. In this case, it seems to me MOM6 requires positive down while HYCOM needs "positive up". Also how about wind direction? Will CMEPS add some rules on wind direction? I think we need to allow component to be "self described", the direction (up/east as positive) can be defined as attributes. Mediator can decide if sign switching is needs for the coupling fields between two components. I am worried as we are using pointers for the coupling fields between fv3 and GOCART, which does not use CMEPS but connectors. Now we have to force fv3 and gocart to change signs because of CMEPS.

**Comment by DeniseWorthen** [comment]
@junwang-noaa I've run the full RT suite and the only tests that are changing are the expected ones (coupled or hafs). 

Edit--removed previous text since I didn't state it correctly. HYCOM uses the fluxes with a (-1) applied. So it also internally defines "positive down" (sfcflux = ~ rad flux - sensible flux - latent flux).

Basically I think the fields exported by a component should be in the sense "positive down" and consistent between components. To me, having ATM export fields to other components with the sign defined for the ATM's use doesn't make sense---these are the fields that are being sent to other components. Within the coupled system, there should be a consistent definition of the direction of the fluxes.

**Comment by junwang-noaa** [comment]
I see. So the proposed change here is that fv3atm only changes signs for those fields sent to ocn (mom6/hycom), not all the fields in its export state such as: 

https://github.com/NOAA-EMC/fv3atm/blob/develop/atmos_model.F90#L1505

**Comment by DeniseWorthen** [comment]
The changes are actually all in the block_data_copy of setup_exportdata:

(https://github.com/DeniseWorthen/fv3atm/blob/32fb333e6eb0df3eabcc4e864ef34edb8d7ac0bd/atmos_model.F90#L2994-L2995)

**Comment by zach1221** [comment]
Ok, I think we're ready to begin here, since 1707 isn't ready. I'll kick things off by adding the jenkins-ci label.

**Comment by FernandoAndrade-NOAA** [comment]
Jenkins ci log is attached, ORTs passed. 
[ufs-weather-model » ort-docker-pipeline » PR-1807 #1 Console [Jenkins].txt](https://github.com/ufs-community/ufs-weather-model/files/12307560/ufs-weather-model.ort-docker-pipeline.PR-1807.1.Console.Jenkins.txt)

**Comment by DeniseWorthen** [comment]
@FernandoAndrade-NOAA What's the issue w/ the hera-bl? I can see that everything appears to have run, but the baselines weren't moved into position so the verification failed.

**Comment by FernandoAndrade-NOAA** [comment]
> @FernandoAndrade-NOAA What's the issue w/ the hera-bl? I can see that everything appears to have run, but the baselines weren't moved into position so the verification failed.

It still seems to be running, but I agree it should have concluded by now. I'm checking for more details on what's going on with the hera-bl run

**Comment by FernandoAndrade-NOAA** [comment]
It doesn't look like baselines were created on hera, rerunning manually.

**Comment by zach1221** [comment]
So, Cheyenne baselines are ready however compile_s2sw_pdlib_intel still has the same scotch loading problem we experienced with 1745. Since 1707 was not merged, I think we should continue working to figure out why the scotch library isn't loading for this compile suite on Cheyenne. @natalie-perlin

**Comment by BrianCurtis-NOAA** [comment]
I brought in changes from the spack-stack PR for WCOSS2/Acorn using nccmp instead of python to fix issues I'm having with the qr intel tests.

**Comment by jkbk2004** [comment]
@natalie-perlin Can you confirm scotch issue on cheyenne? We can make a decision to move this pr once you confirm.

**Comment by zach1221** [comment]
I've talked to Natalie and she believes the scotch issue to be resolved. I will continue testing on Cheyenne

**Comment by zach1221** [comment]
I've got the last RT matching running on Cheyenne currently. I'll update here as soon as it's finished.

**Comment by DeniseWorthen** [comment]
I accidentally pushed an incorrect cmeps branch to this PR, which I then reverted. The CMEPS hash used for testing was 

```
677fa40adf763397b92d372fc870d6f09963c23d ../CMEPS-interface/CMEPS (cmeps_v0.4.1-1407-g677fa40a)
```

which is the current hash in this pr.

**Comment by zach1221** [comment]
Apologies for the delay with this PR. There's a few rrfs_smoke_conus13km qr cases on cheyenne I've been having difficulty getting to match. Hoping to have it figured out soon.

**Comment by zach1221** [comment]
Good morning, @jkbk2004. I was unable to get the four qr cases to match against baselines over the weekend.  I'm making one more attempt now, and if I'm still unable, I think we should skip Cheyenne again and create an issue.
rrfs_smoke_conus13km_hrrr_warm_qr_intel, rrfs_smoke_conus13km_fast_phy32_qr_intel, rrfs_smoke_conus13km_hrrr_warm_qr_gnu, and rrfs_smoke_conus13km_phy32_qr_gnu.

**Comment by DeniseWorthen** [comment]
@zach1221 It looks like the last time we have existing cheyenne tests were for PR https://github.com/ufs-community/ufs-weather-model/pull/1812. Do we know when the qr smoke tests stopped reproducing? ie...have you tried to create/verify baselines for any of the intervening PRs?

**Comment by zach1221** [comment]
Hey, @DeniseWorthen . I'm still trying to pin down when these qr tests started to have issues. I think it started around the time this issue was created [1832](https://github.com/ufs-community/ufs-weather-model/issues/1832). Which I guess makes sense if PR 1812 was the last time these tests matched without issue, but I don't understand why it only fails to match against certain files on Cheyenne only. I think I can get it figured out, but it may take some more time.

**Comment by jkbk2004** [comment]
I agree we need to work on cheyenne adn skip cheyenne for this pr. On my side, I am not able to build even.

**Comment by natalie-perlin** [comment]
@jkbk2004 @zach1221, 
Please let me know what issues still exist on Cheyenne, and how could I reproduce the error. There should be no missing libraries (if the complaint is about the missing libraries).

**Comment by zach1221** [comment]
Hi, @natalie-perlin . Sure, I'll reach out to you if needed. I created this issue [here](https://github.com/ufs-community/ufs-weather-model/issues/1866), to continue working on the problem.

Since it sounds like we're in agreement on the testing being finished for this PR, we can begin the merge process.

**Comment by DeniseWorthen** [comment]
@jkbk2004 If tests being turned off on Cheyenne then we need a push to turn off the non-reproing tests  and then the Cheyenne log posted. As I understand, only a subset of tests is not running on Cheyenne but the baseline otherwise exists and reproduces.

**Comment by zach1221** [comment]
> @jkbk2004 If tests being turned off on Cheyenne then we need a push to turn off the non-reproing tests and then the Cheyenne log posted. As I understand, only a subset of tests is not running on Cheyenne but the baseline otherwise exists and reproduces.

@DeniseWorthen this is correct. I can push the log to this PR that has all the other tests passing. It will just show that those four qr cases failed.

**Comment by DeniseWorthen** [comment]
What we need is 1) turn off the tests in rt.conf for cheyenne. 2) A passing cheyenne.log with those turned off.

**Comment by zach1221** [comment]
> What we need is 1) turn off the tests in rt.conf for cheyenne. 2) A passing cheyenne.log with those turned off.

@DeniseWorthen sure, I can commit the change to rt.conf now, and re-run the tests on Cheyenne following the change.

**Comment by DeniseWorthen** [comment]
@zach1221 At some point you had issues with compiling pdlib cases. How were those resolved?

**Comment by zach1221** [comment]
> @zach1221 At some point you had issues with compiling pdlib cases. How were those resolved?

@DeniseWorthen that was an issue with Scotch loading that Natalie resolved.

File: tests/rt.conf [hunk 1]
```diff
@@ -132,7 +132,7 @@ RUN | rrfs_v1beta                                       |
 RUN | rrfs_v1nssl                                       |                            | baseline |
 RUN | rrfs_v1nssl_nohailnoccn                           |                            | baseline |
 RUN | rrfs_smoke_conus13km_hrrr_warm                    |                            | baseline |
-RUN | rrfs_smoke_conus13km_hrrr_warm_qr                 |                            |          |
+RUN | rrfs_smoke_conus13km_hrrr_warm_qr                 | - cheyenne                            |          |
```
Comment (DeniseWorthen) [inline tests/rt.conf]: Can you line up the vertical bars again for the tests you've turned off? It helps to keep the rt.conf readable.

Comment (zach1221) [reply tests/rt.conf]: @DeniseWorthen Oh, sure. No problem.

**Comment by DeniseWorthen** [comment]
@zach1221 I will update HYCOM and CMEPS. FV3atm is approved, but I can't remember if I have permissions to merge.

**Comment by zach1221** [comment]
@DeniseWorthen Sounds good. No worries on FV3atm. I believe @jkbk2004 has authority to merge.

**Comment by jkbk2004** [comment]
fv3atm pr was merged.

**Comment by zach1221** [comment]
> fv3atm pr was merged.

Thanks, Jong. @DeniseWorthen please update submodule pointer and revert .gitmodules for fv3atm.

Edit: for both cmeps and fv3atm I should say.

**Review by sadeghitabas — APPROVED** [review]

**Review by jkbk2004 — APPROVED** [review]

**Review by Qingfu-Liu — COMMENTED** [review]
The changes look good to me.

**Pull request merged**